### ISSUE TITLE
Pretty-printing for GEM

### ISSUE THREAD

**Issue by miklos1** [firedrakeproject/tsfc#90]
A DAG-aware, nice pretty-printing for GEM would be handy when debugging.

**Comment by miklos1** [comment]
The branch `point-evaluation` contains an old sketch.

**Comment by tj-sun** [comment]
I vaguely remember have done some work on pretty printing in sympy before, might be able to work on this.
Probably can start by overloading _repr_latex_() for jupyter notebook?

**Comment by miklos1** [comment]
LaTeX or similar fancy formatting is not what I had in mind for this issue, but you can create another issue for that.

While the current `__repr__` is okay for simple GEM expressions, for complicated ones it is hardly helpful. I was thinking of something like UFL's `__repr__` _vs_ `__str__`, and also of making this printing DAG-aware, i.e. shared subexpressions should not blow up during printing.

**Comment by tj-sun** [comment]
I see. So we want to implement __str__ in a DAG-aware fashion, which is a nice idea.
I guess later days we can also implement _repr_latex_, in a DAG-aware fashion as well, but yeah that is a separate issue.
Can work on this after finishing COFFEE migration.

**Comment by sv2518** [comment]
I think @narciuch has done something on this using Graphviz. Maybe it's possible to get that to spew a nice string, but google did not give me a straight answer on how.

**Comment by wence-** [comment]
There is also `gem.pprint` which is DAG aware but only textual

**Comment by sv2518** [comment]
Yes true, but I don't think it's nice to read. If pprint is what Miklos was after maybe we can close this?

**Comment by miklos1** [comment]
I started something on DAG-aware pretty printing to ease debugging, which then I abandoned working on as I did not need it that much any more for debugging.  This half-baked version was merged when I had left, that is now `gem.pprint`.  Feel free to do something with graphviz if that helps you.

**Comment by narciuch** [comment]
I've implemented a graph pretty printer on branch tsslac_optimisation, under firedrake/slate/view_gem_dag.py

It can generate DAG aware graphs, simple parent to child trees and DAG unaware full graphs. I've only used it on simpler trees for debugging, so not sure how well it will scale to large trees or if these graph types are useful to others, but its quite an easy tool to work with so shouldn't be too hard to adapt.

The graph can be printed to different formats as provided by Graphviz. I've found the png format quite reader-friendly, but unfortunately haven't seen any that give reader-friendly textual-based output.
